crates/bevy_ecs/src/lib.rs

```diff
@@ -2556,6 +2576,57 @@ mod tests {
         assert_eq!(to_vec(required_z), vec![(b, 0), (c, 1)]);
     }
 
+    #[test]
+    fn update_required_components_in_bundle() {
```

Comment:
Nit for the test: I would also try to test the case where a C->D required component was added later, to ensure that the tracking metadata is properly updated.

Reply:
> You are working on a required components rework. Should I mark my PR to be blocked by yours? Or do you see a quick fix I can smuggle in here?

Mhm, it looks like this bug is still present even in my PR 🤔